subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -500,6 +530,185 @@ static bool dtls_is_peer_addr_valid(struct tls_context *context,
 	return false;
 }
 
+static int tls_session_save(const struct sockaddr *peer_addr,
+			    mbedtls_ssl_session *session)
+{
+	struct tls_session_cache *entry = NULL;
+	size_t session_len;
+	int ret;
+
+	for (int i = 0; i < ARRAY_SIZE(client_cache); i++) {
+		if (client_cache[i].session == NULL) {
+			/* New entry. */
+			if (entry == NULL || entry->session != NULL) {
+				entry = &client_cache[i];
+			}
+		} else {
+			if (peer_addr_cmp(&client_cache[i].peer_addr, peer_addr)) {
+				/* Reuse old entry for given address. */
+				entry = &client_cache[i];
+				break;
+			}
+
+			/* Remember the oldest entry and reuse if needed. */
+			if (entry == NULL ||
+			    (entry->session != NULL &&
+			     entry->timestamp < client_cache[i].timestamp)) {
+				entry = &client_cache[i];
+			}
+		}
+	}
+
+	/* Allocate session and save */
+
+	if (entry->session != NULL) {
+		mbedtls_free(entry->session);
+		entry->session = NULL;
+	}
+
+	(void)mbedtls_ssl_session_save(session, NULL, 0, &session_len);
+
+	entry->session = mbedtls_calloc(1, session_len);
+	if (entry->session == NULL) {
+		NET_ERR("Failed to allocate session buffer.");
+		return -1;
```

Comment:
there are some places returning `-errno`. Wouldn't be better to return `-ENOSPC` here ?

Reply:
Right, some leftoever from prototype code. Changed to ENOMEM which I think suits best here.